File: src/NnCase.Targets.K210/CodeGen/Operators/KPUConv2D.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NnCase.CodeGen;
using NnCase.Evaluation;
using NnCase.Evaluation.Operators;
using NnCase.Runtime;
using NnCase.Targets.K210.IR;
using NnCase.Targets.K210.IR.FakeOperators;
using NnCase.Targets.K210.Runtime.Hardware;
using NnCase.Targets.K210.Runtime.Operators;

namespace NnCase.Targets.K210.CodeGen.Operators
{
    internal static partial class K210Emitters
    {
        private static void RegisterKPUConv2D(CodeGenRegistry registry)
        {
            registry.Add<KPUConv2D>((n, g) =>
            {
                kpu_batchnorm_argument_t[] GetBatchNorm()
                {
                    return n.BatchNorm.Select(x => new kpu_batchnorm_argument_t
                    {
                        norm_mul = (uint)x.Mul,
                        norm_shift = (byte)x.Shift,
                        norm_add = (uint)x.Add
                    }).ToArray();
                }

                unsafe kpu_activate_table_t GetActivation()
                {
                    var table = new kpu_activate_table_t { activate_para = new activate_para_t[16] };
                    for (int i = 0; i < n.Activation.Length; i++)
                    {
                        ref var src = ref n.Activation[i];
                        table.activate_para[i] = new activate_para_t
                        {
                            x_start = (ulong)src.StartX,
                            y_mul = (ushort)src.Mul,
                            shift_number = (byte)src.Shift
                        };

                        if (i < 8)
                            table.activate_para_bias0.result_bias[i] = (byte)src.Add;
                        else
                            table.activate_para_bias1.result_bias[i - 8] = (byte)src.Add;
                    }

                    return table;
                }

                kpu_layer_argument_t GetLayer()
                {
                    var reg = new kpu_layer_argument_t();

                    int oneLoadKernelSize, loadTimes, ocOneTime;
                    if (n.IsDepthwise)
                    {
                        oneLoadKernelSize = (int)n.Weights.Length;
                        loadTimes = 1;
                        ocOneTime = n.KPUOutput.Shape[1];
                    }
                    else
                    {
                        var kernelSize = (int)n.Weights.Length;
                        var filter = KPUShapeUtility.GetKPUFilterSize(n.FilterType);
                        var oneChannelSize = filter * filter * n.Input.Shape[1];
                        var sizeLimit = 30;
                        var oneLoadChannels = Math.Min(n.KPUOutput.Shape[1], (int)Math.Floor(sizeLimit * 1024.0 / oneChannelSize));
                        oneLoadKernelSize = oneChannelSize * oneLoadChannels;
                        loadTimes = (int)Math.Ceiling(n.KPUOutput.Shape[1] / (double)oneLoadChannels);
                        ocOneTime = oneLoadChannels;
                    }

                    var inputLayout = K210Helper.GetRowLayout(n.Input.Shape[3]);
                    var outputLayout = K210Helper.GetRowLayout(n.KPUOutput.Shape[3]);

                    reg.interrupt_enabe = new interrupt_enabe_t
                    {
                        depth_wise_layer = (byte)(n.IsDepthwise ? 1 : 0)
                    };
                    reg.image_addr = new image_addr_t
                    {
                        image_src_addr = K210Helper.GetKpuAddress(g.MemoryRange(n.Input).Start),
                        image_dst_addr = K210Helper.GetKpuAddress(g.MemoryRange(n.KPUOutput).Start)
                    };
                    reg.image_channel_num = new image_channel_num_t
                    {
                        i_ch_num = (ushort)(n.Input.Shape[1] - 1),
                        o_ch_num = (ushort)(n.KPUOutput.Shape[1] - 1),
                        o_ch_num_coef = (ushort)(ocOneTime - 1)
                    };
                    reg.image_size = new image_size_t
                    {
                        i_row_wid = (ushort)(n.Input.Shape[3] - 1),
                        i_col_high = (ushort)(n.Input.Shape[2] - 1),
                        o_row_wid = (ushort)(n.KPUOutput.Shape[3] - 1),
                        o_col_high = (ushort)(n.KPUOutput.Shape[2] - 1)
                    };
                    reg.kernel_pool_type_cfg = new kernel_pool_type_cfg_t
                    {
                        load_para = 1,
                        kernel_type = (byte)n.FilterType,
                        pool_type = (byte)n.PoolType,
                        dma_burst_size = 15,
                        pad_value = n.PadValue
                    };
                    reg.kernel_load_cfg = new kernel_load_cfg_t
                    {
                        load_coor = 1,
                        load_time = (byte)(loadTimes - 1),
                        para_size = (uint)oneLoadKernelSize
                    };
                    reg.kernel_calc_type_cfg = new kernel_calc_type_cfg_t
                    {
                        channel_switch_addr = (ushort)(inputLayout.rowLength * n.Input.Shape[2]),
                        row_switch_addr = (byte)inputLayout.rowLength,
                        coef_group = (byte)inputLayout.groups,
                        load_act = 1
                    };
                    reg.write_back_cfg = new write_back_cfg_t
                    {
                        wb_channel_switch_addr = (ushort)(outputLayout.rowLength * n.KPUOutput.Shape[2]),
                        wb_row_switch_addr = (byte)outputLayout.rowLength,
                        wb_group = (byte)outputLayout.groups
                    };
                    reg.conv_value = new conv_value_t
                    {
                        shr_w = (byte)n.ShiftW,
                        shr_x = (byte)n.ShiftX,
                        arg_w = (uint)n.ArgW,
                        arg_x = (uint)n.ArgX
                    };
                    reg.conv_value2 = new conv_value2_t
                    {
                        arg_add = (ulong)n.ArgAdd
                    };
                    reg.dma_parameter = new dma_parameter_t
                    {
                        channel_byte_num = (ushort)(n.KPUOutput.Shape[3] * n.KPUOutput.Shape[2] - 1),
                        dma_total_byte = (uint)(n.KPUOutput.Shape[3] * n.KPUOutput.Shape[2] * n.KPUOutput.Shape[1] - 1),
                        send_data_out = n.MainMemoryOutput == null ? (byte)0 : (byte)1
                    };

                    return reg;
                }

                return new KPUConv2DOptionsBody
                {
                    Options = new KPUConv2DOptions
                    {
                        MainMemoryOutput = n.MainMemoryOutput == null ? new MemoryRange() : g.MemoryRange(n.MainMemoryOutput),
                        LayerArgument = GetLayer(),
                        BatchNorm = GetBatchNorm(),
                        Activation = GetActivation(),
                        Weights = n.Weights.ToArray()
                    }
                };
            });
        }
    }
}

File: modules/k210/src/codegen/ops/kpu_conv2d.cpp
/* Copyright 2019-2021 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "../module_builder.h"
#include <nncase/ir/ops/k210/runtime_type_utils.h>
#include <nncase/runtime/k210/runtime_op_utility.h>
#include <nncase/runtime/k210/runtime_types.h>

using namespace nncase;
using namespace nncase::codegen;
using namespace nncase::codegen::k210;
using namespace nncase::ir;
using namespace nncase::ir::k210;
using namespace nncase::runtime;
using namespace nncase::runtime::k210;

void k210_module_builder::emit(kpu_conv2d &node) {
    kpu_conv2d_options options{};
    auto &layer = options.layer;
    layer = {};
    int32_t one_load_kernel_size, load_times, oc_one_time;
    if (node.is_depthwise()) {
        one_load_kernel_size = xt::compute_size(node.weights().shape());
        load_times = 1;
        oc_one_time = node.output_channels();
    } else {
        auto filter = get_kpu_filter_size(node.filter_type());
        auto one_channel_size = filter * filter * node.input_channels();
        auto size_limit = 30;
        auto one_load_channels = std::min(
            node.output_channels(),
            (int32_t)std::floor(size_limit * 1024.f / one_channel_size));
        one_load_kernel_size = one_channel_size * one_load_channels;
        load_times = (int32_t)std::ceil(node.output_channels() /
                                        (float)one_load_channels);
        oc_one_time = one_load_channels;
    }

    auto in_layout = get_kpu_row_layout(node.input().shape()[3]);
    auto out_layout = get_kpu_row_layout(node.kpu_output().shape()[3]);

    layer.interrupt_enabe.data.depth_wise_layer = node.is_depthwise();
    layer.image_addr.data.image_src_addr = allocation(node.input()).start / 64;
    layer.image_addr.data.image_dst_addr =
        allocation(node.kpu_output()).start / 64;
    layer.image_channel_num.data.i_ch_num = node.input_channels() - 1;
    layer.image_channel_num.data.o_ch_num = node.output_channels() - 1;
    layer.image_channel_num.data.o_ch_num_coef = oc_one_time - 1;
    layer.image_size.data.i_col_high = (uint16_t)(node.input().shape()[2] - 1);
    layer.image_size.data.i_row_wid = (uint16_t)(node.input().shape()[3] - 1);
    layer.image_size.data.o_col_high =
        (uint16_t)(node.kpu_output().shape()[2] - 1);
    layer.image_size.data.o_row_wid =
        (uint16_t)(node.kpu_output().shape()[3] - 1);
    layer.kernel_pool_type_cfg.data.kernel_type = (uint8_t)node.filter_type();
    layer.kernel_pool_type_cfg.data.load_para = 1;
    layer.kernel_pool_type_cfg.data.pool_type = (uint8_t)node.pool_type();
    layer.kernel_pool_type_cfg.data.dma_burst_size = 15;
    layer.kernel_pool_type_cfg.data.pad_type = 0;
    layer.kernel_pool_type_cfg.data.pad_value = node.pad_value();
    layer.kernel_load_cfg.data.load_coor = 1;
    layer.kernel_load_cfg.data.load_time = (uint8_t)(load_times - 1);
    layer.kernel_load_cfg.data.para_size = (uint32_t)one_load_kernel_size;
    layer.kernel_calc_type_cfg.data.channel_switch_addr =
        (uint16_t)(in_layout.row_len * node.input().shape()[2]);
    layer.kernel_calc_type_cfg.data.row_switch_addr =
        (uint8_t)in_layout.row_len;
    layer.kernel_calc_type_cfg.data.coef_group = (uint8_t)in_layout.groups;
    layer.kernel_calc_type_cfg.data.load_act = 1;
    layer.write_back_cfg.data.wb_channel_switch_addr =
        (uint16_t)(out_layout.row_len * node.kpu_output().shape()[2]);
    layer.write_back_cfg.data.wb_row_switch_addr = (uint8_t)out_layout.row_len;
    layer.write_back_cfg.data.wb_group = (uint8_t)out_layout.groups;
    layer.conv_value.data.arg_x = (uint32_t)node.quant_args().arg_x;
    layer.conv_value.data.shr_x = (uint32_t)node.quant_args().shift_x;
    layer.conv_value.data.arg_w = (uint32_t)node.quant_args().arg_w;
    layer.conv_value.data.shr_w = (uint32_t)node.quant_args().shift_w;
    layer.conv_value2.data.arg_add = (uint64_t)node.quant_args().arg_add;
    layer.dma_parameter.data.channel_byte_num = (uint16_t)(
        node.kpu_output().shape()[3] * node.kpu_output().shape()[2] - 1);
    layer.dma_parameter.data.dma_total_byte =
        (uint32_t)(node.kpu_output().shape()[3] * node.kpu_output().shape()[2] *
                       node.kpu_output().shape()[1] -
                   1);
    layer.dma_parameter.data.send_data_out = node.has_main_mem_output();

    options.weights = allocation(node.weights()).runtime_type();
    options.batch_norm = allocation(node.batch_norm()).runtime_type();
    options.activation = allocation(node.activation()).runtime_type();
    options.main_mem_output =
        node.has_main_mem_output()
            ? allocation(node.main_mem_output()).runtime_type()
            : memory_range{};
    options.batches = (int32_t)node.input().shape()[0];
    text_writer().write(options);
}
